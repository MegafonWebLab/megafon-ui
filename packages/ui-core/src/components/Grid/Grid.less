@import '~styles/base';

.mfui-grid {
    &__container {
        display: flex;
        flex-grow: 1;
    }
    &__container_multi-row {
        flex-wrap: wrap;
    }

    &__container_h-align_right {
        justify-content: flex-end;
    }
    &__container_h-align_center {
        justify-content: center;
    }
    &__container_h-align_between {
        justify-content: space-between;
    }
    &__container_h-align_around {
        justify-content: space-around;
    }

    &__container_v-align_top {
        align-items: flex-start;
    }
    &__container_v-align_bottom {
        align-items: flex-end;
    }
    &__container_v-align_center {
        align-items: center;
    }
    &__container_v-align_baseline {
        align-items: baseline;
    }

    //container gutters
    &__container_gutters-left_large {
        margin-left: -32px;
    }
    &__container_gutters-left_medium {
        margin-left: -20px;
        @media @mobileSM {
            margin-left: -16px;
        }
    }

    //column gutters
    &__column_gutter-left_large {
        padding-left: 32px;
    }
    &__column_gutter-left_medium {
        padding-left: 20px;
        @media @mobileSM {
            padding-left: 16px;
        }
    }
    &__column_gutter-bottom_large {
        padding-bottom: 32px;
    }
    &__column_gutter-bottom_medium {
        padding-bottom: 20px;
        @media @mobileSM {
            padding-bottom: 16px;
        }
    }
}
