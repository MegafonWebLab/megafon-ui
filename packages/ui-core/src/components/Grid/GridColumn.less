@import '~styles/base.less';

.mfui-grid-column {
    box-sizing: border-box;
    &_grow {
        flex-grow: 1;
    }
    &_align_right {
        margin-left: auto;
    }
    &_align_left {
        margin-right: auto;
    }
    &_align_center {
        margin-right: auto;
        margin-left: auto;
    }

    &_flex {
        display: flex;
        flex-shrink: 0;
        flex-grow: 0;
    }

    .generateColumns(all);
    .generateOffsets(all);
    .generateOrder(all-order);

    @media @desktopMU {
        .generateColumns(wide);
        .generateOrder(wide-order);
        .generateOffsets(wide);
    }
    @media @desktopS {
        .generateColumns(desktop);
        .generateOrder(desktop-order);
        .generateOffsets(desktop);
    }
    @media @mobileB {
        .generateColumns(tablet);
        .generateOrder(tablet-order);
        .generateOffsets(tablet);
    }
    @media @mobileSM {
        .generateColumns(mobile);
        .generateOrder(mobile-order);
        .generateOffsets(mobile);
    }
}

.generateColumns(@resolution, @columnCount: 12, @i: 1) when (@i <= @columnCount) {
    &_@{resolution}_@{i} {
        width: (100% / @columnCount) * @i;
    }
    .generateColumns(@resolution, @columnCount, (@i + 1));
}

.generateOrder(@class, @maxOrder:100, @z:0) when (@z <= @maxOrder) {
    &_@{class}_@{z} {
        order: @z;
    }
    .generateOrder(@class, @maxOrder, (@z + 1));
}

.generateOffsets(@resolution, @columnCount: 12, @i: 1) when (@i <= @columnCount) {
    &_offset-@{resolution}_@{i} {
        margin-left: (100% / @columnCount) * @i;
    }
    .generateOffsets(@resolution, @columnCount, (@i + 1));
}
