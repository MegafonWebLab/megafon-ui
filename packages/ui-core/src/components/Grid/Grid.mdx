---
name: Grid
menu: Components
route: components/grid
---

import { Playground, Props } from 'docz';
import Grid from './Grid';
import GridColumn from './GridColumn';
import { itemView } from './Grid.docz';

# Grid

## Properties

<Props of={Grid} />

# GridColumn

## Properties

<Props of={GridColumn} />

## Basic usage

<Playground>
    <Grid>
        <GridColumn>
            <div style={itemView}>Item1</div>
        </GridColumn>
        <GridColumn>
            <div style={itemView}>Item2</div>
        </GridColumn>
        <GridColumn>
            <div style={itemView}>Item3</div>
        </GridColumn>
        <GridColumn>
            <div style={itemView}>Item4</div>
        </GridColumn>
    </Grid>
</Playground>

## Medium gutters with different item's width on all screens usage

<Playground>
    <Grid guttersLeft="medium" guttersBottom="medium">
        <GridColumn all="5"><div style={itemView}>Item1</div></GridColumn>
        <GridColumn all="2"><div style={itemView}>Item2</div></GridColumn>
        <GridColumn all="2"><div style={itemView}>Item3</div></GridColumn>
        <GridColumn all="3"><div style={itemView}>Item4</div></GridColumn>
        <GridColumn all="12"><div style={itemView}>Item7</div></GridColumn>
    </Grid>
</Playground>

## Large gutters with different item's width on all screens usage

<Playground>
    <Grid guttersLeft="large" guttersBottom="large">
        <GridColumn all="5">
            <div style={itemView}>Item1</div>
        </GridColumn>
        <GridColumn all="2">
            <div style={itemView}>Item2</div>
        </GridColumn>
        <GridColumn all="1">
            <div style={itemView}>Item3</div>
        </GridColumn>
        <GridColumn all="3">
            <div style={itemView}>Item4</div>
        </GridColumn>
        <GridColumn all="12">
            <div style={itemView}>Item7</div>
        </GridColumn>
    </Grid>
</Playground>

## Horizontal between alignment usage

<Playground>
    <Grid guttersLeft="large" guttersBottom="large" hAlign="between">
        <GridColumn all="6">
            <div style={itemView}>Item1</div>
        </GridColumn>
        <GridColumn all="5">
            <div style={itemView}>Item2</div>
        </GridColumn>
        <GridColumn all="4">
            <div style={itemView}>Item3</div>
        </GridColumn>
        <GridColumn all="3">
            <div style={itemView}>Item4</div>
        </GridColumn>
        <GridColumn all="2">
            <div style={itemView}>Item5</div>
        </GridColumn>
    </Grid>
</Playground>

## Horizontal around alignment usage

<Playground>
    <Grid guttersLeft="large" guttersBottom="large" hAlign="around">
        <GridColumn all="6">
            <div style={itemView}>Item1</div>
        </GridColumn>
        <GridColumn all="5">
            <div style={itemView}>Item2</div>
        </GridColumn>
        <GridColumn all="4">
            <div style={itemView}>Item3</div>
        </GridColumn>
        <GridColumn all="3">
            <div style={itemView}>Item4</div>
        </GridColumn>
        <GridColumn all="2">
            <div style={itemView}>Item5</div>
        </GridColumn>
    </Grid>
</Playground>

## Horizontal center alignment usage

<Playground>
    <Grid guttersLeft="large" guttersBottom="large" hAlign="center">
        <GridColumn all="6">
            <div style={itemView}>Item1</div>
        </GridColumn>
        <GridColumn all="5">
            <div style={itemView}>Item2</div>
        </GridColumn>
        <GridColumn all="4">
            <div style={itemView}>Item3</div>
        </GridColumn>
        <GridColumn all="3">
            <div style={itemView}>Item4</div>
        </GridColumn>
        <GridColumn all="2">
            <div style={itemView}>Item5</div>
        </GridColumn>
    </Grid>
</Playground>

## Horizontal right alignment usage

<Playground>
    <Grid guttersLeft="large" guttersBottom="large" hAlign="right">
        <GridColumn all="6">
            <div style={itemView}>Item1</div>
        </GridColumn>
        <GridColumn all="5">
            <div style={itemView}>Item2</div>
        </GridColumn>
        <GridColumn all="4">
            <div style={itemView}>Item3</div>
        </GridColumn>
        <GridColumn all="3">
            <div style={itemView}>Item4</div>
        </GridColumn>
        <GridColumn all="2">
            <div style={itemView}>Item5</div>
        </GridColumn>
    </Grid>
</Playground>

## Vertical top alignment usage

<Playground>
    <Grid guttersLeft="medium" hAlign="between" vAlign="top">
        <GridColumn all="4">
            <div style={itemView}>Item1</div>
        </GridColumn>
        <GridColumn all="4">
            <div style={itemView}>
                Item2 <br />
                <br />
            </div>
        </GridColumn>
        <GridColumn all="4">
            <div style={itemView}>
                Item3 <br />
                <br />
                <br />
            </div>
        </GridColumn>
    </Grid>
</Playground>

## Vertical bottom alignment usage

<Playground>
    <Grid guttersLeft="medium" hAlign="between" vAlign="bottom">
        <GridColumn all="4">
            <div style={itemView}>Item1</div>
        </GridColumn>
        <GridColumn all="4">
            <div style={itemView}>
                Item2 <br />
                <br />
            </div>
        </GridColumn>
        <GridColumn all="4">
            <div style={itemView}>
                Item3 <br />
                <br />
                <br />
            </div>
        </GridColumn>
    </Grid>
</Playground>

## Vertical center alignment usage

<Playground>
    <Grid guttersLeft="medium" hAlign="between" vAlign="center">
        <GridColumn all="4">
            <div style={itemView}>Item1</div>
        </GridColumn>
        <GridColumn all="4">
            <div style={itemView}>
                Item2 <br />
                <br />
            </div>
        </GridColumn>
        <GridColumn all="4">
            <div style={itemView}>
                Item3 <br />
                <br />
                <br />
            </div>
        </GridColumn>
    </Grid>
</Playground>

## Vertical baseline alignment usage

<Playground>
    <Grid guttersLeft="medium" hAlign="between" vAlign="baseline">
        <GridColumn all="4">
            <div style={itemView}>Item1</div>
        </GridColumn>
        <GridColumn all="4">
            <div style={itemView}>
                Item2 <br />
                <br />
            </div>
        </GridColumn>
        <GridColumn all="4">
            <div style={itemView}>
                Item3 <br />
                <br />
                <br />
            </div>
        </GridColumn>
    </Grid>
</Playground>

## Without multiRow usage

<Playground>
    <Grid guttersLeft="medium" multiRow={false}>
        <GridColumn all="12">
            <div style={itemView}>Item1</div>
        </GridColumn>
        <GridColumn all="5">
            <div style={itemView}>Item2</div>
        </GridColumn>
        <GridColumn all="5">
            <div style={itemView}>Item3</div>
        </GridColumn>
        <GridColumn all="2">
            <div style={itemView}>Item4</div>
        </GridColumn>
        <GridColumn all="2">
            <div style={itemView}>Item5</div>
        </GridColumn>
    </Grid>
</Playground>

## Different column's width on different screens usage

<Playground>
    <Grid guttersLeft="medium" guttersBottom="medium">
        <GridColumn all="6" tablet="4" mobile="2">
            <div style={itemView}>Item1</div>
        </GridColumn>
        <GridColumn all="5" wide="7" tablet="9" mobile="11">
            <div style={itemView}>Item2</div>
        </GridColumn>
        <GridColumn all="4" tablet="3" desktop="4">
            <div style={itemView}>Item3</div>
        </GridColumn>
        <GridColumn all="3" wide="12">
            <div style={itemView}>Item4</div>
        </GridColumn>
        <GridColumn all="2">
            <div style={itemView}>Item5</div>
        </GridColumn>
    </Grid>
</Playground>

## Change of order items on mobile screens

<Playground>
    <Grid guttersLeft="medium" guttersBottom="medium">
        <GridColumn all="2" mobile="6">
            <div style={itemView}>Item1</div>
        </GridColumn>
        <GridColumn all="8" mobile="12" orderMobile="3">
            <div style={itemView}>Item2</div>
        </GridColumn>
        <GridColumn all="2" mobile="6" orderMobile="2">
            <div style={itemView}>Item3</div>
        </GridColumn>
    </Grid>
</Playground>

## Change of order items on tablet screens

<Playground>
    <Grid guttersLeft="medium" guttersBottom="medium">
        <GridColumn all="2">
            <div style={itemView}>Item1</div>
        </GridColumn>
        <GridColumn all="8" mobile="12" orderTablet="3">
            <div style={itemView}>Item2</div>
        </GridColumn>
        <GridColumn all="2" mobile="6" orderTablet="2">
            <div style={itemView}>Item3</div>
        </GridColumn>
    </Grid>
</Playground>

## Change of order items on different screens

<Playground>
    <Grid guttersLeft="medium" guttersBottom="medium">
        <GridColumn all="2" mobile="6" orderWide="3">
            <div style={itemView}>Item1</div>
        </GridColumn>
        <GridColumn all="8" mobile="12" orderTablet="3" orderWide="2">
            <div style={itemView}>Item2</div>
        </GridColumn>
        <GridColumn all="2" mobile="6" orderTablet="2" orderWide="1">
            <div style={itemView}>Item3</div>
        </GridColumn>
    </Grid>
</Playground>

## Offset usage

<Playground>
    <Grid guttersLeft="medium" guttersBottom="medium">
        <GridColumn all="2" offsetAll="3"><div style={itemView}>Item1</div></GridColumn>
        <GridColumn all="5" offsetAll="1"><div style={itemView}>Item2</div></GridColumn>
        <GridColumn all="3" offsetAll="2"><div style={itemView}>Item3</div></GridColumn>
        <GridColumn all="4" offsetAll="1"><div style={itemView}>Item4</div></GridColumn>
        <GridColumn all="4" offsetAll="1"><div style={itemView}>Item5</div></GridColumn>
        <GridColumn all="3" offsetAll="1"><div style={itemView}>Item5</div></GridColumn>
    </Grid>
</Playground>
