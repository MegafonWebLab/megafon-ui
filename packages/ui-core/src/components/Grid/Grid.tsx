import * as React from 'react';
import * as PropTypes from 'prop-types';
import './Grid.less';
import cnCreate from 'utils/cnCreate';
import { IGridColumn } from './GridColumn';

export interface IGridProps {
    /** Выравнивание всех колонок по горизонтали */
    hAlign?: 'left' | 'right' | 'center' | 'between' | 'around';
    /** Выравнивание всех колонок по вертикали */
    vAlign?: 'top' | 'bottom' | 'center' | 'baseline';
    /** Отступ слева от колонки */
    guttersLeft?: 'large' | 'medium';
    /** Отступ снизу от колонки */
    guttersBottom?: 'large' | 'medium';
    /** Перенос столбцов в несколько строк */
    multiRow?: boolean;
    /** Дополнительный класс корневого элемента */
    className?: string;
    children: Array<React.ReactElement<IGridColumn>> | React.ReactElement<IGridColumn>;
}

const cn = cnCreate('mfui-beta-grid');
const Grid: React.FC<IGridProps> = ({
    children,
    guttersLeft,
    guttersBottom,
    multiRow = true,
    hAlign,
    vAlign,
    className,
}) => (
    <div className={cn([className])}>
        <div
            className={cn('container', {
                'multi-row': multiRow,
                'h-align': hAlign,
                'v-align': vAlign,
                'gutters-left': guttersLeft,
                'gutters-bottom': guttersBottom,
            })}>
            {React.Children.map(children, (child: React.ReactElement<IGridColumn>) =>
                React.cloneElement(child, {
                    className: cn(
                        'column',
                        {
                            'gutter-left': guttersLeft,
                            'gutter-bottom': guttersBottom,
                        },
                        child.props.className
                    ),
                })
            )}
        </div>
    </div>
);

Grid.propTypes = {
    hAlign: PropTypes.oneOf(['left', 'right', 'center', 'between', 'around']),
    vAlign: PropTypes.oneOf(['top', 'bottom', 'center', 'baseline']),
    guttersLeft: PropTypes.oneOf(['large', 'medium']),
    guttersBottom: PropTypes.oneOf(['large', 'medium']),
    multiRow: PropTypes.bool,
    className: PropTypes.string,
    children: PropTypes.oneOfType([
        PropTypes.arrayOf(PropTypes.element.isRequired),
        PropTypes.element.isRequired,
    ]).isRequired,
};

export default Grid;
