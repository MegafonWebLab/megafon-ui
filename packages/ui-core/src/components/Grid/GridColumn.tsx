import * as React from 'react';
import * as PropTypes from 'prop-types';
import './GridColumn.less';
import cnCreate from 'utils/cnCreate';

const GridSizeValues = PropTypes.oneOf(['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12']);
export type TGridSizeValues = '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9' | '10' | '11' | '12';

export interface IGridColumn {
    /** Size of columns on wide screens */
    wide?: TGridSizeValues;
    /** Size of columns on desktop screens */
    desktop?: TGridSizeValues;
    /** Size of columns on tablet screens */
    tablet?: TGridSizeValues;
    /** Size of columns on mobile screens */
    mobile?: TGridSizeValues;
    /** Size of columns on all screens */
    all?: TGridSizeValues;

    /** Order on all screens */
    orderAll?: string;
    /** Order on wide screens */
    orderWide?: string;
    /** Order on desktop screens */
    orderDesktop?: string;
    /** Order on tablet screens */
    orderTablet?: string;
    /** Order on mobile screens */
    orderMobile?: string;

    /** Size of offset on wide screens */
    offsetWide?: TGridSizeValues;
    /** Size of offset on desktop screens */
    offsetDesktop?: TGridSizeValues;
    /** Size of offset on tablet screens */
    offsetTablet?: TGridSizeValues;
    /** Size of offset on mobile screens */
    offsetMobile?: TGridSizeValues;
    /** Size of offset on all screens */
    offsetAll?: TGridSizeValues;

    /** Custom alignment of column */
    align?: 'right' | 'left' | 'center';
    /** Column flex grow */
    grow?: boolean;
    /** Column as flex container */
    flex?: boolean;
    /** Custom class name */
    className?: string;
    children: JSX.Element[] | Element[] | JSX.Element | string | Element;
}

const cn = cnCreate('mfui-grid-column');
class GridColumn extends React.Component<IGridColumn, {}> {
    static propTypes = {
        wide: GridSizeValues,
        desktop: GridSizeValues,
        tablet: GridSizeValues,
        mobile: GridSizeValues,
        all: GridSizeValues,

        orderAll: PropTypes.string,
        orderWide: PropTypes.string,
        orderDesktop: PropTypes.string,
        orderTablet: PropTypes.string,
        orderMobile: PropTypes.string,

        offsetWide: GridSizeValues,
        offsetDesktop: GridSizeValues,
        offsetTablet: GridSizeValues,
        offsetMobile: GridSizeValues,
        offsetAll: GridSizeValues,

        align: PropTypes.oneOf(['right', 'left', 'center']),
        grow: PropTypes.bool,
        flex: PropTypes.bool,
        className: PropTypes.string,
        children: PropTypes.oneOfType([
            PropTypes.arrayOf(PropTypes.element),
            PropTypes.element,
            PropTypes.string,
        ]),
    };

    static defaultProps = {
        all: '12',
        flex: false,
        grow: false,
        orderWide: '0',
        orderDesktop: '0',
        orderTablet: '0',
        orderMobile: '0',
        orderAll: '0',
    };

    render() {
        const {
            all,
            wide,
            desktop,
            tablet,
            mobile,
            orderAll,
            orderWide,
            orderDesktop,
            orderTablet,
            orderMobile,
            offsetAll,
            offsetWide,
            offsetDesktop,
            offsetTablet,
            offsetMobile,
            className,
            flex,
            grow,
            align,
            children,
        } = this.props;

        return (
            <div
                className={cn(
                    '',
                    {
                        flex,
                        grow,
                        align,
                        all,
                        wide,
                        desktop,
                        tablet,
                        mobile,
                        'all-order': orderAll,
                        'wide-order': orderWide,
                        'desktop-order': orderDesktop,
                        'tablet-order': orderTablet,
                        'mobile-order': orderMobile,
                        'offset-all': offsetAll,
                        'offset-wide': offsetWide,
                        'offset-desktop': offsetDesktop,
                        'offset-tablet': offsetTablet,
                        'offset-mobile': offsetMobile,
                    },
                    className
                )}
            >
                {children}
            </div>
        );
    }
}

export default GridColumn;
