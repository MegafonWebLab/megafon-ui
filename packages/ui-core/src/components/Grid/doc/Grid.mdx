---
name: Grid
menu: Компоненты
route: components/grid
---

import { Props } from 'docz';
import Grid from '../Grid';
import GridColumn from '../GridColumn';
import Example from './Grid.example.mdx';

# Grid

<CopyToClipboardBox text="import { Grid, GridColumn } from '@megafon/ui-core';" />

<DoczTabs>
    <Example />
    <div>
        <Props of={Grid} />
        <h2>GridColumn</h2>
        <Props of={GridColumn} />
    </div>
</DoczTabs>
