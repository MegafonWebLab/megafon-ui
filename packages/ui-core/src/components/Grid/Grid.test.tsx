import * as React from 'react';
import { shallow } from 'enzyme';
import Grid from './Grid';
import GridColumn from './GridColumn';

describe('<Grid />', () => {
    it('it renders Grid without props', () => {
        const wrapper = shallow(
            <Grid>
                <GridColumn>child1</GridColumn>
                <GridColumn>child2</GridColumn>
            </Grid>
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('it renders Grid with props', () => {
        const wrapper = shallow(
            <Grid
                hAlign="right"
                vAlign="top"
                guttersLeft="large"
                guttersBottom="medium"
                multiRow={false}
                className={'custom-class-name'}
            >
                <GridColumn>child1</GridColumn>
                <GridColumn>child2</GridColumn>
            </Grid>
        );
        expect(wrapper).toMatchSnapshot();
    });
});
