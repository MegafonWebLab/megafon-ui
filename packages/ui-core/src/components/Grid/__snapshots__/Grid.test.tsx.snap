// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Grid /> it renders Grid with props 1`] = `
<div
  className="mfui-beta-grid custom-class-name"
>
  <div
    className="mfui-beta-grid__container mfui-beta-grid__container_h-align_right mfui-beta-grid__container_v-align_top mfui-beta-grid__container_gutters-left_large mfui-beta-grid__container_gutters-bottom_medium"
  >
    <GridColumn
      className="mfui-beta-grid__column mfui-beta-grid__column_gutter-left_large mfui-beta-grid__column_gutter-bottom_medium"
      key=".0"
    >
      child1
    </GridColumn>
    <GridColumn
      className="mfui-beta-grid__column mfui-beta-grid__column_gutter-left_large mfui-beta-grid__column_gutter-bottom_medium"
      key=".1"
    >
      child2
    </GridColumn>
  </div>
</div>
`;

exports[`<Grid /> it renders Grid without props 1`] = `
<div
  className="mfui-beta-grid"
>
  <div
    className="mfui-beta-grid__container mfui-beta-grid__container_multi-row"
  >
    <GridColumn
      className="mfui-beta-grid__column"
      key=".0"
    >
      child1
    </GridColumn>
    <GridColumn
      className="mfui-beta-grid__column"
      key=".1"
    >
      child2
    </GridColumn>
  </div>
</div>
`;
